src/query/functions/binary/base.go

```diff
@@ -29,76 +29,76 @@ import (
 	"github.com/m3db/m3/src/query/parser"
 )
 
-type logicalOp struct {
+type baseOp struct {
 	OperatorType string
-	LNode        parser.NodeID
-	RNode        parser.NodeID
-	makeBlock    makeBlockFn
-	Matching     *VectorMatching
-	ReturnBool   bool
+	processFunc  processFunc
+	params       NodeParams
+}
+
+// NodeParams describes the types of nodes used
+// for binary operations
+type NodeParams struct {
+	LNode, RNode         parser.NodeID
+	LIsScalar, RIsScalar bool
+	ReturnBool           bool
+	VectorMatching       *VectorMatching
 }
 
 // OpType for the operator
-func (o logicalOp) OpType() string {
+func (o baseOp) OpType() string {
 	return o.OperatorType
 }
 
 // String representation
-func (o logicalOp) String() string {
-	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.LNode, o.RNode)
+func (o baseOp) String() string {
+	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.params.LNode, o.params.RNode)
 }
 
 // Node creates an execution node
-func (o logicalOp) Node(controller *transform.Controller, _ transform.Options) transform.OpNode {
-	return &logicalNode{
+func (o baseOp) Node(controller *transform.Controller, _ transform.Options) transform.OpNode {
+	return &baseNode{
+		op:         o,
+		process:    o.processFunc,
 		controller: controller,
 		cache:      transform.NewBlockCache(),
-		op:         o,
 	}
 }
 
-type logicalNode struct {
-	op         logicalOp
-	controller *transform.Controller
-	cache      *transform.BlockCache
-	mu         sync.Mutex
-}
-
-type makeBlockFn func(
-	logicalNode *logicalNode,
-	lIter, rIter block.StepIter,
-) (block.Block, error)
-
-// NewLogicalOp creates a new logical operation
-func NewLogicalOp(
+// NewOp creates a new binary operation
+func NewOp(
 	opType string,
-	lNode parser.NodeID,
-	rNode parser.NodeID,
-	matching *VectorMatching,
+	params NodeParams,
 ) (parser.Params, error) {
-	var makeBlock makeBlockFn
-	switch opType {
-	case AndType:
-		makeBlock = makeAndBlock
-	case OrType:
-		makeBlock = makeOrBlock
-	case UnlessType:
-		makeBlock = makeUnlessBlock
-	default:
-		return logicalOp{}, fmt.Errorf("operator not supported: %s", opType)
+	fn, ok := buildLogicalFunction(opType, params)
+	if !ok {
+		fn, ok = buildArithmeticFunction(opType, params)
+		if !ok {
+			fn, ok = buildComparisonFunction(opType, params)
+			if !ok {
```

Comment:
Anyway to clean this code up a bit? Looks odd having 3 nested `if` statements.

Reply:
I tried a few things but unfortunately there's not many ways to add sugar to functions that return 2 values. Best alternative I came up with was something like 
```
fn, ok = buildArithmeticFunction(opType, params)
		if !ok {
			fn, ok = buildBinaryFunction(opType, params)
                        if !ok {
                          return error
```
and have buildBinaryFunction try building an Arithmetic then a Comparison function, but that doesn't achieve much except for kicking the can down the road